File: zig/test/fast_sort.c
#include <stdlib.h>
#include <stdio.h>
#include <time.h>
#include <string.h>


void sort_fast(int *arr, int arr_size) {
    if (arr_size == 1) {
        return;
    }
    if (arr_size > 2) {
        int l_arr_size = 0;
        int r_arr_size = 0;
        if (arr_size % 2 == 0) {
            // int l_arr[arr_size / 2];
            // int r_arr[arr_size / 2];

            int *l_arr = malloc(sizeof(int) * (arr_size / 2));
            int *r_arr = malloc(sizeof(int) * (arr_size / 2));

            memcpy(l_arr, arr, (arr_size / 2) * 4);
            memcpy(r_arr, arr + arr_size / 2, (arr_size / 2) * 4);
            sort_fast(l_arr, arr_size / 2);
            sort_fast(r_arr, arr_size / 2);
            l_arr_size = r_arr_size = arr_size / 2;
            // int fin_arr[arr_size];
            int *fin_arr = malloc(sizeof(int) * arr_size);
            int l_i = 0;
            int r_i = 0;
            for(int i = 0; i < arr_size; i++) {
                if (l_i == l_arr_size && r_i < r_arr_size) {
                    fin_arr[i] = r_arr[r_i++];
                    continue;
                } 
                if (r_i == r_arr_size && l_i < l_arr_size) {
                    fin_arr[i] = l_arr[l_i++];
                    continue;
                }
                if (l_arr[l_i] > r_arr[r_i]) {
                    fin_arr[i] = r_arr[r_i++];
                    continue;
                } else {
                    fin_arr[i] = l_arr[l_i++];
                    continue;
                }
            }
            memcpy(arr, fin_arr, arr_size * 4);
            free(l_arr);
            free(r_arr);
            free(fin_arr);
        } else {
            // int l_arr[arr_size / 2 + 1];
            // int r_arr[arr_size / 2];
            int *l_arr = malloc(sizeof(int) * (arr_size / 2 + 1));
            int *r_arr = malloc(sizeof(int) * (arr_size / 2));
            memcpy(l_arr, arr, (arr_size / 2 + 1) * 4);
            memcpy(r_arr, arr + arr_size / 2 + 1, (arr_size / 2) * 4);
            sort_fast(l_arr, arr_size / 2 + 1);
            sort_fast(r_arr, arr_size / 2);
            l_arr_size = arr_size / 2 + 1;
            r_arr_size = arr_size / 2;

            //int fin_arr[arr_size];
            int *fin_arr = malloc(sizeof(int) * arr_size);
            int l_i = 0;
            int r_i = 0;
            for(int i = 0; i < arr_size; i++) {
                if (l_i == l_arr_size && r_i < r_arr_size) {
                    fin_arr[i] = r_arr[r_i++];
                    continue;
                } 
                if (r_i == r_arr_size && l_i < l_arr_size) {
                    fin_arr[i] = l_arr[l_i++];
                    continue;
                }
                if (l_arr[l_i] > r_arr[r_i]) {
                    fin_arr[i] = r_arr[r_i++];
                    continue;
                } else {
                    fin_arr[i] = l_arr[l_i++];
                    continue;
                }
            }

            memcpy(arr, fin_arr, arr_size * 4);
            free(l_arr);
            free(r_arr);
            free(fin_arr);
        }
    } else if (arr_size == 2) {
        if (arr[0] > arr[1]) {
            int tmp = arr[0];
            arr[0] = arr[1];
            arr[1] = tmp;
            return;
        }
    }
}

#define count 1000000

int main() {
    srand(time(NULL));

    int arr[count] = {0};
    for(int i = 0; i < count; i++) {
        arr[i] = rand();
    }

    // for(int i = 0; i < count; i++) {
    //     printf("%d\n", arr[i]);
    // }

    // printf("========\n");
    printf("Start\n");
    sort_fast(arr, count);
    printf("End\n");
    //
    // for(int i = 0; i < count; i++) {
    //     printf("%d\n", arr[i]);
    // }

    return 0;
}

File: zig/test/fast_sort.zig
const std = @import("std");

pub fn fast_sort(allocator: std.mem.Allocator, arr: []u32) !void {
    if (arr.len <= 1) return;
    if (arr.len > 2) {
        if (arr.len % 2 == 0) {
            const left_arr = try allocator.alloc(u32, arr.len / 2);
            const right_arr = try allocator.alloc(u32, arr.len / 2);
            @memcpy(left_arr, arr[0..(arr.len / 2)]);
            @memcpy(right_arr, arr[(arr.len / 2)..arr.len]);
            try fast_sort(allocator, left_arr);
            try fast_sort(allocator, right_arr);

            var fin_arr = try allocator.alloc(u32, arr.len);
            @memcpy(fin_arr, arr);
            var l_i: usize = 0;
            var r_i: usize = 0;
            var i: usize = 0;
            while(i < fin_arr.len) : (i += 1) {
                if (l_i == left_arr.len) {
                    fin_arr[i] = right_arr[r_i];
                    r_i += 1;
                    continue;
                } 
                if (r_i == right_arr.len) {
                    fin_arr[i] = left_arr[l_i];
                    l_i += 1;
                    continue;
                }
                if (right_arr[r_i] > left_arr[l_i]) {
                    fin_arr[i] = left_arr[l_i];
                    l_i += 1;
                } else {
                    fin_arr[i] = right_arr[r_i];
                    r_i += 1;
                }
            }
            allocator.free(left_arr);
            allocator.free(right_arr);
            @memcpy(arr, fin_arr);
            allocator.free(fin_arr);

        } else {
            const left_arr = try allocator.alloc(u32, arr.len / 2 + 1);
            const right_arr = try allocator.alloc(u32, arr.len / 2);
            @memcpy(left_arr, arr[0..(arr.len / 2 + 1)]);
            @memcpy(right_arr, arr[(arr.len / 2 + 1)..arr.len]);
            try fast_sort(allocator, left_arr);
            try fast_sort(allocator, right_arr);

            var fin_arr = try allocator.alloc(u32, arr.len);
            @memcpy(fin_arr, arr);
            var l_i: usize = 0;
            var r_i: usize = 0;
            var i: usize = 0;
            while(i < fin_arr.len) : (i += 1) {
                if (l_i == left_arr.len) {
                    fin_arr[i] = right_arr[r_i];
                    r_i += 1;
                    continue;
                } 
                if (r_i == right_arr.len) {
                    fin_arr[i] = left_arr[l_i];
                    l_i += 1;
                    continue;
                }
                if (right_arr[r_i] > left_arr[l_i]) {
                    fin_arr[i] = left_arr[l_i];
                    l_i += 1;
                } else {
                    fin_arr[i] = right_arr[r_i];
                    r_i += 1;
                }
            }
            allocator.free(left_arr);
            allocator.free(right_arr);
            @memcpy(arr, fin_arr);
            allocator.free(fin_arr);
        }
    } else {
        if (arr[0] > arr[1]) {
            const tmp = arr[0];
            arr[0] = arr[1];
            arr[1] = tmp;
            return;
        }
    }
}

const count = 1000000;

pub fn main() !void {
    //var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    var arena = std.heap.ArenaAllocator.init(std.heap.c_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();
    var arr: [count]u32 = .{0} ** count;
    for(0..count) |i| {
        arr[i] = std.crypto.random.int(u32);
    }

    // for(&arr) |i| {
    //     std.debug.print("{d}\n", .{i});
    // }
    //
    // std.debug.print("========\n", .{});

    std.debug.print("Start\n", .{});
    try fast_sort(allocator, &arr);
    std.debug.print("End\n", .{});

    // for(&arr) |i| {
    //     std.debug.print("{d}\n", .{i});
    // }
}


